odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,472 @@
+
+        # Get byte range and set start
+        byte_range = field[keys['byte_range']].split('-')
+        byte_start = int(byte_range[0]) - 1
+        byte_end = int(byte_range[-1]) - 1
+        new_field['offset_bytes'] = byte_start
+
+        # Data type: transform to Numpy format and get shape from its
+        # itemsize and the byte range
+        dtype = dtype_map[field[keys['dtype']]]
+        byte_size = byte_end - byte_start + 1
+
+        if hasattr(dtype, 'itemsize') and byte_size % dtype.itemsize:
+            raise ValueError(
+                'in row {}: byte range {} not a multiple of itemsize {} '
+                'of the data type {}.'
+                ''.format(row + 1, field[keys['byte_range']], dtype.itemsize,
+                          field[keys['dtype']]))
+
+        new_field['dtype'] = dtype
+        new_field['size_bytes'] = byte_size
+        # Assuming 1d arrangement of multiple elements
+        # TODO: find way to handle 2d fields
+        if nend:
+            new_field['dshape'] = (nend - nstart + 1,)
+        elif hasattr(dtype, 'itemsize'):
+            new_field['dshape'] = (byte_size / dtype.itemsize,)
+        else:
+            new_field['dshape'] = (1,)
+
+        conv_list.append(new_field)
+
+    return tuple(conv_list)
+
+
+class FileReaderUncompressedBinary(object):
+
+    """Reader for uncompressed binary files including a header."""
+
+    def __init__(self, file, header_fields):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        file : file-like or str
+            Stream or filename from which to read the data. The stream
+            is allowed to be already opened in 'rb' mode.
+        header_fields : sequence of dicts
+            Definition of the fields in the header (per row), each
+            containing key-value pairs for the following keys:
+
+            - ``'name'`` : Label for the field.
+            - ``'offset_bytes'`` : Start of the field in bytes.
+            - ``'size_bytes'`` : Size of the field in bytes.
+            - ``'dtype'`` : Data type in Numpy- or Numpy-readable format.
+            - ``'dshape'`` (optional) : The array of values is reshaped to
+              this shape.
+            - ``'description'`` : A human-readable description of the field.
+        """
+        try:
+            f = open(file, 'rb')
+        except TypeError:
+            f = file
+
+        if f.mode != 'rb':
+            raise ValueError("`file` must be opened in 'rb' mode, but mode "
+                             "'is {}'".format(f.mode))
+
+        self.file = f
+
+        # Initialize some attributes to default values, plus the header
+        self.data_shape = -1  # Makes reshape a no-op
+        self.header_bytes = None
+        self.header_fields = header_fields
+        self.data = None
+        if self.header_fields:
+            self.header = self.read_header()
+        else:
+            self.header = None
+
+    @classmethod
```

Comment:
Yes, I think there is no elegant way around this. What the constructor needs to do is to use a raw specification `header_bytes` and `dtype` and bypass the usual constructor completely. I see no way of doing that by calling `__init__` with some special case of arguments. The problem is that the size of the header is calculated on the fly since it depends on an entry in the header that gives the size of the extended header (MRC). Such a thing seems to be used quite often (e.g. also in the `.npy` format).